### ISSUE TITLE
CPU multiprocess conflict with GPU

### ISSUE THREAD

**Issue by simone-mastrogiovanni** [simone-mastrogiovanni/CINECA_hackhaton_icarus#2]
icarogw uses [bilby](https://lscsoft.docs.ligo.org/bilby/). Bilby is a python package for Bayesian inference. Icarogw interacts with bilby in the following way

- icarogw implements a bilby likelihood model [here](https://github.com/simone-mastrogiovanni/icarogw/blob/main/icarogw/likelihood.py#LL7C1-L7C6).
- The likelihood model does all the computations either using numpy or cupy
- The likelihood model [returns](https://github.com/simone-mastrogiovanni/icarogw/blob/main/icarogw/likelihood.py#L85) a CPU float value to bilby. 
- Bilby do whatever it needs to do.

The problem is the following. Bilby can work by multithreading CPUs. This is done with [multiprocessing](https://docs.python.org/3/library/multiprocessing.html). We want the GPU computations of the likelihood of icarogw to work with the multiprocessing implementation of Bilby. Currently if you try to run [this script](https://github.com/simone-mastrogiovanni/CINECA_hackhaton_icarus/blob/main/profiling_GPU_bilby_problem.py) you get the following error

```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "profiling_GPU_bilby_problem.py", line 81, in <module>
    res = bilby.core.sampler.run_sampler(likelihood, priors=prior_dict, label='bilby_problem', outdir='bilby_runs', sampler='dynesty',npool=2)
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/site-packages/bilby/core/sampler/__init__.py", line 234, in run_sampler
    result = sampler.run_sampler()
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/site-packages/bilby/core/sampler/base_sampler.py", line 96, in wrapped
    output = method(self, *args, **kwargs)
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/site-packages/bilby/core/sampler/dynesty.py", line 517, in run_sampler
    out = self._run_external_sampler_with_checkpointing()
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/site-packages/bilby/core/sampler/dynesty.py", line 635, in _run_external_sampler_with_checkpointing
    self.sampler.run_nested(**sampler_kwargs)
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/site-packages/dynesty/sampler.py", line 1017, in run_nested
    for it, results in enumerate(
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/site-packages/dynesty/sampler.py", line 840, in sample
    u, v, logl, nc = self._new_point(loglstar_new)
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/site-packages/dynesty/sampler.py", line 421, in _new_point
    u, v, logl, nc, blob = self._get_point_value(loglstar)
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/site-packages/dynesty/sampler.py", line 404, in _get_point_value
    self._fill_queue(loglstar)
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/site-packages/dynesty/sampler.py", line 397, in _fill_queue
    self.queue = list(mapper(evolve_point, args))
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/m100/home/userexternal/smastrog/.conda/envs/icarogw/lib/python3.8/multiprocessing/pool.py", line 771, in get
    raise self._value
cupy_backends.cuda.api.runtime.CUDARuntimeError: cudaErrorInitializationError: initialization error
```

There are no errors when you try to run 1 GPU&1 CPU or N-CPU&0GPU

**Comment by nouiz** [comment]
It would be great if we can get more information about the initialization error.
Is there a way to have the CPU only process created before we initialize the GPU?
That allowed to some issues that this one remind me.

**Comment by simone-mastrogiovanni** [comment]
Some followup on this issue from the work that I did/I am doing. I searched on stackoverflow ([here](https://stackoverflow.com/questions/54808148/cupy-get-error-in-multithread-pool-if-gpu-already-used)) and apparently the solutions are two

1.  Initialize cupy only after the multiprocessing has been initialized (as we discussed).
2.  Use the `multiprocessing.set_start_method('forkserver', force=True)` or `multiprocessing.set_start_method('spawn', force=True)`

About Solution 1:

The first solution involves modifying the python package [bilby](https://lscsoft.docs.ligo.org/bilby/). Because the pool setup is initialized in [this line](https://github.com/simone-mastrogiovanni/CINECA_hackhaton_icarus/blob/main/profiling_GPU_bilby_problem.py#L80) in an internal `bilby` function. So the only way that I see to go for Solution 1 is to modify bilby (Not preferred).

About Solution 2:

I both tried inserting `multiprocessing.set_start_method('forkserver', force=True)`  and `multiprocessing.set_start_method('spawn', force=True)`. I had to modify a bit the scripts, e.g. you should not have hdf5 pointers open. After removing all the files etc, I managed to start the process but I get the following warning.

```
raise RuntimeError('''
RuntimeError: 
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.

```
And then
```
Traceback (most recent call last):
  File "profiling_GPU_bilby_problem.py", line 87, in <module>
    res  = bilby.core.sampler.run_sampler(likelihood, priors=prior_dict, label='bilby_problem', outdir='bilby_runs', sampler='dynamic_dynesty',npool=2)
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/site-packages/bilby/core/sampler/__init__.py", line 234, in run_sampler
    result = sampler.run_sampler()
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/site-packages/bilby/core/sampler/base_sampler.py", line 96, in wrapped
    output = method(self, *args, **kwargs)
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/site-packages/bilby/core/sampler/dynesty.py", line 491, in run_sampler
    self._setup_pool()
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/site-packages/bilby/core/sampler/dynesty.py", line 548, in _setup_pool
    super(Dynesty, self)._setup_pool()
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/site-packages/bilby/core/sampler/base_sampler.py", line 749, in _setup_pool
    self.pool = multiprocessing.Pool(
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/multiprocessing/context.py", line 119, in Pool
    return Pool(processes, initializer, initargs, maxtasksperchild,
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/multiprocessing/pool.py", line 212, in __init__
    self._repopulate_pool()
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/multiprocessing/pool.py", line 303, in _repopulate_pool
    return self._repopulate_pool_static(self._ctx, self.Process,
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/multiprocessing/pool.py", line 326, in _repopulate_pool_static
    w.start()
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/multiprocessing/context.py", line 291, in _Popen
    return Popen(process_obj)
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/multiprocessing/popen_forkserver.py", line 35, in __init__
    super().__init__(process_obj)
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/storage/users/mastrogiovanni/anaconda3/envs/icarogw2.0/lib/python3.8/multiprocessing/popen_forkserver.py", line 58, in _launch
    f.write(buf.getbuffer())
BrokenPipeError: [Errno 32] Broken pipe

```

**Comment by simone-mastrogiovanni** [comment]
Update on this. We managed to solve the issue. Apparently, cupy can work with multiprocessing if it is imported and used after the pools are initialized.

The approach I used to enable cupy without touching bilby is to enable cupy at a given ellapsed time in this [function](https://github.com/simone-mastrogiovanni/icarogw/blob/main/icarogw/likelihood.py#L44). I will try to do it in a cleaner way and commit a patch.

Below a screenshot of the two pools using the same GPU.

![image](https://github.com/simone-mastrogiovanni/CINECA_hackhaton_icarus/assets/62158795/fd6350f3-b1dd-4e67-b1f9-8b0a7596bd6b)
